cf-keycrypt/cf-keycrypt.c

```diff
@@ -0,0 +1,504 @@
+    "cf-keycrypt offers a simple way to encrypt or decrypt files using keys "
+    "generated by cf-key. CFEngine uses asymmetric cryptography, and "
+    "cf-keycrypt allows you to encrypt a file using a public key file. "
+    "The encrypted file can only be decrypted on the host with the "
+    "corresponding private key. Original author: Jon Henrik Bjornstad "
+    "<jonhenrik@cfengineers.net>";
+
+static const struct option OPTIONS[] =
+{
+    {"help",        no_argument,        0, 'h'},
+    {"manpage",     no_argument,        0, 'M'},
+    {"encrypt",     no_argument,        0, 'e'},
+    {"decrypt",     no_argument,        0, 'd'},
+    {"key",         required_argument,  0, 'k'},
+    {"host",        required_argument,  0, 'H'},
+    {"output",      required_argument,  0, 'o'},
+    {NULL,          0,                  0, '\0'}
+};
+
+static const char *const HINTS[] =
+{
+    "Print the help message",
+    "Print the man page",
+    "Encrypt file",
+    "Decrypt file",
+    "Use key file",
+    "Encrypt for host (get key from lastseen database)",
+    "Output file",
+    NULL
+};
+
+void *get_in_addr(struct sockaddr *sa)
+{
+    if (sa->sa_family == AF_INET)
+    {
+        return &(((struct sockaddr_in*)sa)->sin_addr);
+    }
+    return &(((struct sockaddr_in6*)sa)->sin6_addr);
+}
+
+int file_exist(const char *filename)
+{
+    struct stat buffer;
+    return stat(filename, &buffer) == 0;
+}
+
+char *get_host_pubkey(const char *host)
+{
+    char *buffer = (char *) malloc(BUFSIZE *sizeof(char));
+    char hash[CF_HOSTKEY_STRING_SIZE];
+    char ipaddress[BUFSIZE];
+    struct addrinfo *result;
+    struct addrinfo *res;
+    bool found = false;
+
+    int error = getaddrinfo(host, NULL, NULL, &result);
+    if (error != 0)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to get IP from host (getaddrinfo: %s)",
+            gai_strerror(error));
+        return NULL;
+    }
+
+    for(res = result; res != NULL && !found; res = res->ai_next)
+    {
+        inet_ntop(res->ai_family, get_in_addr((struct sockaddr *)res->ai_addr), ipaddress, sizeof(ipaddress));
+        if ((strcmp(ipaddress, "127.0.0.1") == 0) || (strcmp(ipaddress, "::1") == 0))
+        {
+            found = true;
+            snprintf(buffer, BUFSIZE * sizeof(char), "%s/ppkeys/localhost.pub", WORKDIR);
+            return buffer;
+        }
+        found = Address2Hostkey(hash, sizeof(hash), ipaddress);
+    }
+    if (found)
+    {
+        snprintf(buffer, BUFSIZE * sizeof(char), "%s/ppkeys/root-%s.pub", WORKDIR, hash);
+        freeaddrinfo(result);
+        return buffer;
+    }
+    else
+    {
+        for(res = result; res != NULL; res = res->ai_next)
+        {
+            inet_ntop(
+                res->ai_family,
+                get_in_addr((struct sockaddr *)res->ai_addr),
+                ipaddress,
+                sizeof(ipaddress));
+            snprintf(
+                buffer, BUFSIZE * sizeof(char), "%s/ppkeys/root-%s.pub",
```

Comment:
We don't really use this kind of line breaking elsewhere in our code, unless it is almost unavoidable. Not a blocker, just keep in mind for next time.